csrc/multidevice/executor.cpp

```diff
@@ -98,10 +98,29 @@ MultiDeviceExecutor::MultiDeviceExecutor(
       copyFusionAndChangeOutputs(completeFusion(), vals_to_allocate_);
 }
 
-void MultiDeviceExecutor::postKernel(SegmentedGroup* group) {
+void MultiDeviceExecutor::postKernel(
+    SegmentedGroup* group,
+    const LaunchParams& l_params) {
   if (!should_run_.at(group)) {
     return;
   }
+
+  if (!params_.use_fusion_executor_cache) {
+    // erase from group's input extent corresponding to grid and block dims. Set
+    // them through launchparams.
+    std::vector<Val*> group_input_vals;
+    for (auto input : group->inputs()) {
+      if (input->isA<NamedScalar>()) {
```

Comment:
```suggestion
      if (input->isScalar()) {
```
Is there a reason to treat named ones specially?

Reply:
yes, this is on purpose.
First of all, let me recall that we don't support scalar inputs for now (just for simplicity). Here, the NamedScalar correspond to the parallel dimension "gridDim" and "blockDim", which are not "real" inputs of the Fusion, but rather symbolic extent inserted by the ReductionScheduler that need to be binded to concrete values at runtime. This is why we treat them in a special path. That is why we "`continue`": we don't know need to treat them as actual inputs, and the `FusionExecutor` how to bind those `Val` with the concrete values contained in the `launch_params`. Let me recall the two asserts in the code ensuring all of the above:
```
NVF_ERROR(input->as<NamedScalar>()->getParallelDim().has_value());
NVF_ERROR(l_params.hasDim(
    input->as<NamedScalar>()->getParallelDim().value()));
```

Now, Let me point out that this change is only necessary for `SchedulingMode::ReductionOnly`, in which we use `FusionExecutor` and not `FusionExecutorCache`. So I would call that as well the "additional bookkeeping" annoying but necessary in this case, as we already talked about multiple times.
This remark also holds for the changes in `kernel_cache.cpp` (see [this comment](https://github.com/NVIDIA/Fuser/pull/1853#discussion_r1534688986))

So, this change is exactly the kind of hack around that I don't know how to do better when using FusionExecutor. This falls back to our previous discussions.

Reply:
> we don't support scalar inputs for now (just for simplicity)

Can you clarify what support is missing? AFAICS from [this test](https://github.com/NVIDIA/Fuser/blob/5817da992a052d6005fd22621f4140fbc3bcfa6c/tests/cpp/test_scalar_hoisting.cpp#L347-L349), FusionExecutor does support scalar inputs. So I doubt simplicity is the real reason.

Reply:
I think nothing fundamental is missing, this is just a simplifying assumption we make (among others). But someone could go and add this support :) It's one of our TODO. I think it should be pretty straightforward but we haven't done it and it's not a priority in my opinion.

Reply:
(when I say "we don't support" I'm talking exclusively about the multidevice module)

Reply:
> it's same code complexity but covers more cases (allowing non-grid/block-dim scalars).

IIUC you suggest to move the asserts into the "if" condition, which I did.

> FusionExecutor already have problems taking grid/block dims and therefore we have to erase them?

It is not that FusionExecutor has a problem taking those inputs. It's more that `blockDim` and `gridDim` need to be passed to  `FusionExecutor::runFusion` (at least to match the expected number of inputs!), while being of a different nature than the other inputs because they come from the scheduler's `launch_params`.

I'm trying to clarify the situation as much as I can here:
1. First, the scheduler is used on the Fusion and includes some symbolic extents `gridDim` and `blockDim` in the tensor.
2. Then, the segmenter is used, and it treats these symbolic extents as group's I/O, and add them in the group's `input_vals`. So the group's inputs are `{input 1, ..., input n, blockDim, gridDim}`. But we see here that we are mixing two inputs of different natures: real inputs and inputs coming from the scheduler's `launch_params`. Let me point out here that this situation was not encountered before because this is the first time we use the segmenter AFTER scheduling. 
Whether the segmenter should add `gridDim` and `blockDim` as group's input is a question in itself, I think it makes sense in terms of dependency, but it requires a special handling later on (as the one I implemented here), so that's debatable. I chose not to change the scheduler and to handle it manually.
3. Inside `MultiDeviceExecutor::postKernel`, the segmented group is compiled into a kernel, and we need to create the kernel's concrete input called `group_input_IValues` to pass it to `FusionExecutor::runFusion`. We populate `group_input_IValues` by retrieving concrete values from the global context of previously stored values. However, `blockDim` and `gridDim` should not be retrieved from global context per say but rather from the`launch_params` (passed from the scheduler to our MultiDeviceExecutor). The question is: "what to do with `blockDim` and `gridDim`". We could either:
  a. retrieve the concrete values manually from `launch_params` and add then to `group_input_IValues`
  b. remove them from group's input, pass the `launch_params` to `FusionExecutor::runFusion` and rely on that function for handling it (which is already implemented).
I chose solution b. because a. is redundant with `FusionExecutor::runFusion` already handling the `launch_params`.

I hope this detailed explanation makes things clearer. Let me know if it's still unclear, or if you have a better suggestion.

Reply:
> So the group's inputs are {input 1, ..., input n, blockDim, gridDim}.

Maybe the real problem is the scheduler adding `blockDim` and `gridDim` as inputs. They are not inputs but intrinsic variables automatically defined in CUDA. What do you think, @jacobhinkle ?

Reply:
Thanks -- that clarifies lots of my questions and is the level of details I wanted to hear. I put some of your words into the code comment for posterity.

Reply:
> First, the scheduler is used on the Fusion and includes some symbolic extents gridDim and blockDim in the tensor.

How so? I thought we do only inter-device scheduling before segmentation, so gridDim and blockDim shouldn't be there at least conceptually.

Reply:
> Maybe the real problem is the segmenter adding `blockDim` and `gridDim` as inputs. They are not inputs but intrinsic variables automatically defined in CUDA. What do you think, @jacobhinkle ?

From the CPU perspective they are actually launch parameters, so user can control it, in this sense it is an input.

> How so? I thought we do only inter-device scheduling before segmentation, so gridDim and blockDim shouldn't be there at least conceptually.

No, doing intra-device scheduling is precisely the goal of this pr. See the test and `SchedulingMode::ReductionOnly`

Reply:
> SchedulingMode::Auto implements flow 2.

yes



> SchedulingMode::ReductionOnly implements flow 1

not exactly, I would say

Inter AND INTRA-device scheduling => inter-device only segmentation (which means we just segment out the resharding ops)  => FE::runFusion these scheduled intra-device segments (without further segmentation).

Reply:
> AND INTRA

Is this necessary? Inter-device segmentation needs only DID schedules. If we can move the intra-device scheduling after segmentation, I think it will save us some unnecessary complexity (e.g. this one)

Reply:
In my opinion, this is necessary, and the steps are done in the right order, which is:
1) Fusion definition
2) scheduling
3) execution

the user should be able to schedule a Fusion (as done in the test), and then rely on the MultiDeviceExecutor to automatically lower/compile/execute it. Here what you call the "inter-device segmentation" is part of the lowering process, because we need to insert the necessary communications (what I mean: the segmentation step here doesn't make any decision on the parallelization strategy, it just does what is necessary to execute the scheduled Fusion).

does it make sense?

> I think it will save us some unnecessary complexity

I don't think this is an unnecessary complexity, especially if we want to make the segmenter, scheduler, and FusionExecutor a bit more modulable than what it currently is, and understand what implicit assumption they rely on.

This complexity is the type of additional "bookkeeping" necessary when not using FusionExecutorCache.

Reply:
Sorry if my question wasn't clear. I was asking about intra-device scheduling. Why does that have to happen before inter-device segmentation? 

(And I understood we are talking about a flow that will eventually go away.)

Reply:
I understood your question correctly, and thought I addressed it, sorry if I haven't.
I think that we shouldn't treat differently inter and intra device scheduling. Both belongs to the same step, and should happen strictly before lowering and execution.
 
If that's still unclear maybe it's better to have this high-level discussion during a call?

Reply:
Why do you think it should be done in the order you describe?
The "avoided complexity" I can see relates to implicit assumption made in the implementation of the segmenter/scheduler

Reply:
> (And I understood we are talking about a flow that will eventually go away.)

Sorry I forgot this question.

I think that it's pretty unclear at this point. cf our previous discussion. It will depend on how the segmenter, scheduler, segmenter will be refactored. Here, despite adding implementation complexity, I make the effort to implement two paths (FusionExecutor and FusionExecutorCache) to better anticipate future change, as you advocate during our previous discussions

Reply:
> From the CPU perspective they are actually launch parameters, so user can control it, in this sense it is an input.

But if you leave them as segment inputs, since they are never "produced", that segment would never be ready to run, right?

Reply:
Since this is a temporary workaround, I don't have any strong preference. I think this is just a matter of:

> Maybe the real problem is the segmenter adding blockDim and gridDim as inputs. They are not inputs but intrinsic variables automatically defined in CUDA.

Reply:
I suspect this issue is also automatically fixed:

https://github.com/NVIDIA/Fuser/pull/1853/files/40160a398a71fda65a7af48485430510c92386d9#r1538108135